docs/design-documents/drivers/serial/serial.md

```diff
@@ -91,27 +91,23 @@ class BufferedSerial : private SerialBase, public FileHandle, private NonCopyabl
 
 **API description**
 
-These are the new constructors for `BufferedSerial`. The first constructor creates the buffers dynamically.
+These are the constructors for `BufferedSerial`. The first constructor statically allocates the receive and transmit buffers with a configurable size of MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE and MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE respectively.
 
 ```C
     /** Create a BufferedSerial port, connected to the specified transmit and receive pins, with a particular baud rate.
      *  @param tx Transmit pin
      *  @param rx Receive pin
      *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
-     *  @param rx_buffer_size The size of the receive buffer (optional, defaults to MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE)
-     *  @param tx_buffer_size The size of the transmit buffer (optional, defaults to MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE)
      *
      *  @note
      *    Either tx or rx may be specified as NC if unused
      */
     BufferedSerial(PinName tx,
                    PinName rx,
-                   int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE,
-                   size_t rx_buffer_size = MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE,
-                   size_t tx_buffer_size = MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE);
+                   int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE);
 ```
 
-The second constructor takes buffer pointers as arguments.
+If MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE and MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE are set to zero, then another constructor is provided to allow the application to pass different buffers based on the intended usage of the serial instance. 
```

Comment:
The declaration is missing.

@kjbracey-arm is there some C++ magic that would allow us to have both of them at the same time without wasting memory for default buffers and not using malloc?

```
serial_console = BifferedSerial(tx, rx); // Default buffers
modem = BufferedSerail(tx, rx, tx, tx_size, rx, rx_size); // Default buffers are not used or allocated in this instance
```

Reply:
Yes, I've been discussing with Hugues.

My current proposal is to have it be a template so that `BufferedSerial<500,200>` gives you built-in static buffers of the specified size. Like `CircularBuffer` currently does.

All the implementation can be in a non-templated `BufferedSerialBase` that has external buffers, avoiding bloat. The template only needs to have the constructor passthrough and the buffer definition.

But we can borrow a trick from `std::span` to unify the dynamic and non-dynamic forms. At the minute we have `CircularBuffer<N>` - we can add a specialisation for `CircularBuffer<-1>` (giving `-1` some nice name, like `span's` `dynamic_extent`). 

For `std::span` (and I would propose `CircularBuffer`), that's the default template parameter, so `std::span<char>` means `std::span<char, dynamic_extent>`. The dynamic and non-dynamic forms are really different types, overloaded by the template specialisation, but they look the same in use.

For `BufferedSerial`, the default would presumably be the configured sizes built into it, rather than dynamic, but we could allow `BufferedSerial<external_buffer>(pins, rx_buffer, rx_buffer_len, tx_buffer, tx_buffer_len)`.

Reply:
@bulislaw The declaration is following (unchanged from previous version).

I am not happy with `If MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE and MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE are set to zero, then another constructor is provided` and it seems that @kjbracey-arm 's solution does not need it so I will remove the mention of `MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE and MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE`.

Kevin - I haven't understood how `BufferedSerial<dynamic>(RX, TX, rx_buf, rx_buf_len, tx_buf, tx_buf_len)` avoids using/allocating the internal buffers.

Reply:
I was thinking of a distinct specialisation - `BufferedSerial<dynamic>` literally doesn't have the buffers.

    template <size_t TXBufSize = MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE,
              size_t RxBufSize = MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE>
    class BufferedSerial : public BufferedSerialBase {
    public:
        BufferedSerial(PinName tx, PinName rx) :
                BufferedSerialBase(tx, rx, rx_buf_, sizeof rx_buf_, tx_buf_, sizeof tx_buf_) { }
    private:
        uint8_t tx_buf_[TxBufSize];
        uint8_t rx_buf_[RxBufSize];
    }

    template<>
    class BufferedSerial<dynamic> : public BufferedSerialBase {
    public:
        // would use `using` to pick up base constructors, but ARMC5 doesn't support
         BufferedSerial(PinName tx, PinName rx, void *tx_buf, size_t tx_size, void *rx_buf, size_t rx_size) : BufferedSerialBase(tx, rx, tx_buf, tx_size, rx_buf, rx_size) { }
    }

Note that specialisation is actually `BufferedSerial<dynamic, MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE>` because of the defaulted second parameter - I'm assuming people would rather type `BufferedSerial<dynamic>` than `BufferedSerial<dynamic, dynamic>`. Doing better than this would triple the size and render it un-doxygenable.

I thought about the internal dynamic allocation case, but it made my brain hurt. Let's avoid it.

Reply:
That's good UX but now i'm worried about the footprint :(